### PR TITLE
Remove #ifs in System.ComponentModel.TypeConverter

### LINKED ISSUE #62212: Remove #ifs in System.ComponentModel.TypeConverter
Related to #39174.

We have some left over #ifs that weren't removed as part of #18862.

See:

https://github.com/dotnet/runtime/blob/bb597e2f6b4f47f0f3058b12f16a83e4f3b5e6ba/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectTypeDescriptionProvider.cs#L78-L87

https://github.com/dotnet/runtime/blob/bb597e2f6b4f47f0f3058b12f16a83e4f3b5e6ba/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectTypeDescriptionProvider.cs#L652-L661

and one non-behavior instance that can just be cleaned up:

https://github.com/dotnet/runtime/blob/bb597e2f6b4f47f0f3058b12f16a83e4f3b5e6ba/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/StandardCommands.cs#L665-L667

For the first two, we should remove the `#if` and enable the code it is protecting. These are left-over from pre-standard2.0 days when some APIs weren't available. As part of enabling this code we should add unit tests for the behavior.

For the last one, the whole code chunk can be removed since it is not needed.

### PR DESCRIPTION
**Pull request by pedrobsaila**
Fixes #62212

### CODE REVIEW

File: src/libraries/System.ComponentModel.TypeConverter/src/Properties/InternalsVisibleTo.cs [hunk 1]
```diff
@@ -0,0 +1,6 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("System.ComponentModel.TypeConverter.Tests, PublicKey=00240000048000009400000006020000002400005253413100040000010001004b86c4cb78549b34bab61a3b1800e23bfeb5b3ec390074041536a7e3cbd97f5f04cf0f857155a8928eaa29ebfd11cfbbad3ba70efea7bda3226c6a8d370a4cd303f714486b6ebc225985a638471e6ef571cc92a4613c00b8fa65d61ccee0cbe5f36330c9a01f4183559f1bef24cc2917c6d913e3a541333a1d05d9bed22b38cb")]
```
Comment (tarekgh) [inline src/libraries/System.ComponentModel.TypeConverter/src/Properties/InternalsVisibleTo.cs:6]: Can we avoid this? It is possible if we use the reflection in the test code.  Linkers can trim stuff which can cause issues with this technique.

**Review by tarekgh — APPROVED** [review]
@pedrobsaila thanks a lot for helping with this issue.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectTypeDescriptionProvider.cs [hunk 2]
```diff
@@ -79,10 +79,6 @@ private static Type[] InitializeSkipInterfaceAttributeList()
         {
             return new Type[]
             {
-#if FEATURE_SKIP_INTERFACE
```
Comment (eerhardt) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectTypeDescriptionProvider.cs:82]: This is not what I had intended when I wrote the issue.

> For the first two, we should remove the #if and enable the code it is protecting. These are left-over from pre-standard2.0 days when some APIs weren't available. As part of enabling this code we should add unit tests for the behavior.

To be explicit, I meant:

```diff
- #if FEATURE_SKIP_INTERFACE
                typeof(System.Runtime.InteropServices.GuidAttribute),
                typeof(System.Runtime.InteropServices.InterfaceTypeAttribute)
- #endif
```

Which means this code would now look like:

```C#
        private static Type[] InitializeSkipInterfaceAttributeList()
        {
            return new Type[]
            {
                typeof(System.Runtime.InteropServices.GuidAttribute),
                typeof(System.Runtime.InteropServices.InterfaceTypeAttribute)
                typeof(System.Runtime.InteropServices.ComVisibleAttribute),
            };
        }
```

See https://github.com/dotnet/corefx/pull/12977 for how we removed these `#if`s previously.

The added unit tests should tests should fail when using the current product code, and pass with the new changes. They should be testing that `GuidAttribute` and `InterfaceTypeAttribute` are skipped correctly.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectTypeDescriptionProvider.cs [hunk 3]
```diff
@@ -649,16 +645,6 @@ protected internal override IExtenderProvider[] GetExtenderProviders(object inst
                 {
                     return GetExtenders(extenderList.GetExtenderProviders(), instance, cache);
                 }
-#if FEATURE_COMPONENT_COLLECTION
```
Comment (eerhardt) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectTypeDescriptionProvider.cs:652]: Same as above.

**Review by eerhardt — DISMISSED** [review]

**Comment by eerhardt** [comment]
It looks like some trimming tests are broken:

```
 Unhandled exception. System.NotImplementedException: The method or operation is not implemented.
     at DbConnectionStringBuilderTrimmingTests.TestSite.get_Container() in D:\a\_work\1\s\artifacts\bin\trimmingTests\projects\System.Data.Common.TrimmingTests\DbConnectionStringBuilder\win-x64\DbConnectionStringBuilder.cs:line 139
     at System.ComponentModel.ReflectTypeDescriptionProvider.GetExtenderProviders(Object)
     at System.ComponentModel.ReflectTypeDescriptionProvider.GetExtendedProperties(Object)
     at System.ComponentModel.TypeDescriptor.TypeDescriptionNode.DefaultExtendedTypeDescriptor.System.ComponentModel.ICustomTypeDescriptor.GetProperties()
     at System.ComponentModel.TypeDescriptor.GetPropertiesImpl(Object, Attribute[], Boolean, Boolean)
     at System.Data.Common.DbConnectionStringBuilder.GetProperties(Hashtable propertyDescriptors)
     at DbConnectionStringBuilderTrimmingTests.Program.Main(String[]) in D:\a\_work\1\s\artifacts\bin\trimmingTests\projects\System.Data.Common.TrimmingTests\DbConnectionStringBuilder\win-x64\DbConnectionStringBuilder.cs:line 90
```

Can you look into this?

File: src/libraries/System.ComponentModel.TypeConverter/tests/ComponentResourceManagerTests.cs [hunk 4]
```diff
@@ -164,10 +164,12 @@ public void ApplyResources_IComponentWithDesignModeSite_Success()
 
         private class TestSite : ISite
         {
+            private TestContainer _testContainer = new TestContainer();
+
             public bool DesignMode { get; set; }
 
             public IComponent Component => throw new NotImplementedException();
-            public IContainer Container => throw new NotImplementedException();
+            public IContainer Container => _testContainer;
```
Comment (eerhardt) [inline src/libraries/System.ComponentModel.TypeConverter/tests/ComponentResourceManagerTests.cs:172]: Maybe we should just return null here.

```suggestion
            public IContainer Container => null;
```

File: src/libraries/System.Data.Common/tests/TrimmingTests/DbConnectionStringBuilder.cs [hunk 5]
```diff
@@ -134,14 +135,28 @@ public Hashtable GetProperties2()
 
     class TestSite : INestedSite
     {
+        private TestContainer _testContainer = new TestContainer();
+
         public string FullName => null;
         public IComponent Component => throw new NotImplementedException();
-        public IContainer Container => throw new NotImplementedException();
+        public IContainer Container => _testContainer;
```
Comment (eerhardt) [inline src/libraries/System.Data.Common/tests/TrimmingTests/DbConnectionStringBuilder.cs:142]: No need to make it more complicated than it has to be.

```suggestion
        public IContainer Container => null;
```

File: src/libraries/System.ComponentModel.TypeConverter/tests/ReflectTypeDescriptionProviderTests.cs [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+            testComponent.Disposed += (object obj, EventArgs args) => { };
+            IExtenderProvider[] extenderProviders = provider.GetType().GetMethod("GetExtenderProviders", BindingFlags.Instance | BindingFlags.NonPublic)
+                .Invoke(provider, new[] { testComponent }) as IExtenderProvider[];
+            IExtenderProvider result = Assert.Single(extenderProviders);
+            Assert.IsType<TestExtenderProvider>(result);
+        }
+
+        private static object CreateReflectTypeDescriptionProviderInstance()
+        {
+            Assembly srcAsssemby = Assembly.Load(Assembly.GetExecutingAssembly().GetReferencedAssemblies().FirstOrDefault(a => a.FullName.StartsWith("System.ComponentModel.TypeConverter")));
```
Comment (eerhardt) [inline src/libraries/System.ComponentModel.TypeConverter/tests/ReflectTypeDescriptionProviderTests.cs:58]: We should be able to test these scenarios using public APIs. The `ReflectTypeDescriptionProvider` is the default one, so you should be able to use the normal APIs to do this work. Let's only use reflection as a last resort.

**Review by eerhardt — COMMENTED** [review]
Let's remove the reflection usages in the new tests and instead use public APIs.

**Review by eerhardt — APPROVED** [review]
LGTM. Thanks for the contribution here @pedrobsaila!

**Comment by dreddy-work** [comment]
@pedrobsaila @eerhardt , this change is causing some test issues in WinForms repo (linked issue above). Any pointers on how we can solve this?

**Pull request merged**